firedrake/output.py

```diff
@@ -1,562 +1,481 @@
-        if coords_cgN and function_cgN:
-            family = 'CG'
-            project_coords = not coords_p1
-            project_function = not function_p1
+    :arg coordinates: The coordinates defining the mesh.
+    :returns: A tuple of ``(connectivity, offsets, types)``
+        :class:`OFunction`\s.
+    """
+    V = coordinates.function_space()
+    mesh = V.ufl_domain().topology
+    cell = mesh.ufl_cell()
+    values = V.cell_node_map().values.astype("int32")
+    # Non-simplex cells need reordering
+    # Connectivity of bottom cell in extruded mesh
+    if cells[cell] == VTK_QUADRILATERAL:
+        # Quad is
+        #
+        # 1--3    3--2
+        # |  | -> |  |
+        # 0--2    0--1
+        values = values[:, [0, 2, 3, 1]]
+    elif cells[cell] == VTK_WEDGE:
+        # Wedge is
+        #
+        #    5          5
+        #   /|\        /|\
+        #  / | \      / | \
+        # 1----3     3----4
+        # |  4 | ->  |  2 |
+        # | /\ |     | /\ |
+        # |/  \|     |/  \|
+        # 0----2     0----1
+        values = values[:, [0, 2, 4, 1, 3, 5]]
+    elif cells[cell] == VTK_HEXAHEDRON:
+        # Hexahedron is
+        #
+        #   5----7      7----6
+        #  /|   /|     /|   /|
+        # 4----6 | -> 4----5 |
+        # | 1--|-3    | 3--|-2
+        # |/   |/     |/   |/
+        # 0----2      0----1
+        values = values[:, [0, 2, 3, 1, 4, 6, 7, 5]]
+    elif cells.get(cell) is None:
+        raise ValueError("Unhandled cell type %r" % cell)
+
+    # Repeat up the column
+    if mesh.layers is None:
+        cell_layers = 1
+    else:
+        cell_layers = mesh.layers - 1
+
+    connectivity = numpy.repeat(values, cell_layers, axis=0)
+
+    if is_continuous(V):
+        scale = 1
+    else:
+        scale = cell.num_vertices()
+
+    offsets = numpy.arange(cell_layers) * scale
+
+    # Add offsets going up the column
+    num_cells = mesh.cell_set.size
+    connectivity += numpy.tile(offsets.reshape(-1, 1), (num_cells, 1))
+
+    num_cells *= cell_layers
+
+    connectivity = connectivity.flatten()
+
+    offsets = numpy.arange(start=cell.num_vertices(),
+                           stop=cell.num_vertices() * (num_cells + 1),
+                           step=cell.num_vertices(),
+                           dtype="int32")
+    cell_types = numpy.full(num_cells, cells[cell], dtype="uint8")
+    return (OFunction(connectivity, "connectivity", None),
+            OFunction(offsets, "offsets", None),
+            OFunction(cell_types, "types", None))
+
+
+def get_byte_order(dtype):
+    import sys
+    native = {"little": "LittleEndian", "big": "BigEndian"}[sys.byteorder]
+    return {"=": native,
+            "|": "LittleEndian",
+            "<": "LittleEndian",
+            ">": "BigEndian"}[dtype.byteorder]
+
+
+def write_array(f, ofunction):
+    array = ofunction.array
+    numpy.uint32(array.nbytes).tofile(f)
+    if get_byte_order(array.dtype) == "BigEndian":
+        array = array.byteswap()
```

Comment:
Yes.  I could probably just remove this entirely and hard-code little endian, given that I don't know of anyone running firedrake on ppc, but this is reasonably safe I think.